schema/schema.hh

```diff
@@ -826,6 +827,14 @@ public:
     static void set_default_partitioner(const sstring& class_name, unsigned ignore_msb = 0);
     const dht::i_partitioner& get_partitioner() const;
     const dht::sharder& get_sharder() const;
+
+    // Returns a pointer to the table if the local database has a table which this object references by id().
+    // The table pointer is not guaranteed to be stable, schema_ptr doesn't keep the table alive.
+    replica::table* maybe_table() const;
+
+    // Like maybe_table() but throws replica::no_such_column_family if the table is not set.
+    replica::table& table() const;
+
```

Comment:
It's really like to avoid this. A schema isn't a table.

Unfortunately shard_of deeply embedded in contexts where we don't have any global knowledge, like mutation::shard_of(). So it will be a long project to adjust all shard queries to refer to global state instead of schema dependent state. But I don't see a way to avoid it now.

Reply:
Don't we already have a concept of a sharder? Can't we use it to abstract how shard assignment works?

Reply:
> Don't we already have a concept of a sharder? Can't we use it to abstract how shard assignment works?

Yes, that's what this series does. What do you propose to change?

Reply:
You can ease the pain by reframing it like this. schema_ptr represents a table schema version. Each table schema version refers to a particular physical table, by its id(). table() is a materialization of that relationship. 

It's similar to how a prepared select statement stores a schema_ptr instead of table name.

Reply:
It's say it's even wrong (but less wrong) for a schema to store the table_id. A table has a current schema but a schema that once belonged to a table cannot be said to own it. It's possible to have schemas not related to any table. For relational databases it's more obvious since a table schema and a query schema are both tabular.

The current schema::get_sharder() says: "this is the algorithm and parameters used to determine s key's shard".

After the change, it no longer refers to an algorithm and parameters, because there is lots of mutable state indirectly referred to. The schema is no longer a value, it contains references.

So, we should work to get rid of get_sharder(). It's lots of work so it will have to happen later.

Reply:
> It's say it's even wrong (but less wrong) for a schema to store the table_id. A table has a current schema but a schema that once belonged to a table cannot be said to own it. It's possible to have schemas not related to any table. For relational databases it's more obvious since a table schema and a query schema are both tabular.

I get your point about having schema which doesn't have to be associated with a table, but that's not what our current schema object is. It is a table schema, which is associated with a table. It doesn't own the table, it refers to it. There is `database::find_column_family(schema_ptr)` which allows one to lookup the table from schema_ptr. If we remove table_id from it, we need to pass around not schema_ptr but a new object:

```
struct table_schema {
    table_id id;
    raw_schema_ptr s;
}
```

This refactor is orthogonal to the argument, because then we would be arguing about having table() inside this new table_schema object.

Reply:
Regardless of the table() thing, I agree we should phase out uses of `schema::get_sharder()`.

Reply:
> > It's say it's even wrong (but less wrong) for a schema to store the table_id. A table has a current schema but a schema that once belonged to a table cannot be said to own it. It's possible to have schemas not related to any table. For relational databases it's more obvious since a table schema and a query schema are both tabular.
> 
> I get your point about having schema which doesn't have to be associated with a table, but that's not what our current schema object is. It is a table schema, which is associated with a table. It doesn't own the table, it refers to it. There is `database::find_column_family(schema_ptr)` which allows one to lookup the table from schema_ptr. If we remove table_id from it, we need to pass around not schema_ptr but a new object:
> 
> ```
> struct table_schema {
>     table_id id;
>     raw_schema_ptr s;
> }
> ```
> 
> This refactor is orthogonal to the argument, because then we would be arguing about having table() inside this new table_schema object.

Maybe I can rephrase the point about table_id as "schema{,_ptr} is misnamed". But then it becomes an exercise in arguing semantics, let's not do that.

So the conclusion is to drop schema::get_sharder(). Probably it will be easier once we wean tablets off all multishard things. Or maybe not, and we just need to decorate effective_replication_map with the sharder and change all those APIs to get an effective_replication_map_ptr (or a replica-side equivalent). That will also serve to prevent from the mapping changing under our feet.